code/modules/mechs/mech_damage.dm

```diff
@@ -72,15 +92,39 @@
 	return def_zone //Careful with effects, mechs shouldn't be stunned
 
 /mob/living/exosuit/getarmor(def_zone, type)
-	if(body?.armor_plate)
-		var/body_armor = body.armor_plate?.armor.getRating(type)
-		if(body_armor) return body_armor
+	if(!def_zone)
+		def_zone = ran_zone()
+	var/obj/item/mech_component/hit_zone = zoneToComponent(def_zone)
+	var/armor = hit_zone?.armor.getRating(type)
+	if(armor)
+		return armor
 	return 0
 
 /mob/living/exosuit/updatehealth()
 	if(body) maxHealth = body.mech_health
 	health = maxHealth - (getFireLoss() + getBruteLoss())
 
+/mob/living/exosuit/damage_through_armor(damage, damagetype, def_zone, attack_flag, armor_divisor, used_weapon, sharp, edge, wounding_multiplier, list/dmg_types, return_continuation, dir_mult)
+	var/obj/item/mech_component/comp = zoneToComponent(def_zone)
+	var/armor_def = comp.armor.getRating(attack_flag)
+	var/deflect_chance = ((comp.shielding + armor_def)*0.5) - (armor_divisor*5)
```

Comment:
We actually have a math function for deflection using our penetration system somewhere in the dev channels, I can get it for you if you'd like.

Reply:
Sure, does it actually ricochet the bullet ala like with walls? That'd be ideal I think

Reply:
Couldn't find the formula, but in short:

in `/obj/item/projectile/bullet/check_penetrate(...)` you'll find `var/percentile_blocked = block_damage(blocked_damage, A)`

The same block_damage proc should be used by the mech's armor. This reduces the bullet's damage by the specified amount (taking AP into account), and returns a % of the damage left as a number.

At this point, check if the number returned is above 0.5 (50%). If it is, deal damage as normal, check for penetration etc.

If it is below 0.5 however, the bullet was deflected. Deal only half of the bullet's damage as NON-sharp NON-edge, then repurpose the following proc's formula: 
https://github.com/Humonitarian/CEV-Eris/blob/d9ca2807d5653c1d90e03f8320cd9bfddf359563/code/modules/clothing/spacesuits/rig/rig_pieces.dm#L53

Except without checking for chance, at this point the bullets are guaranteed to ricochet.


This same ricochet calculation will likely be added to RIGs as well with the new damage system. (It was removed during the rework as the numbers no longer made sense, and it was unnecessary RNG.)

Reply:
Hm, I think space changed it to use code that makes it actually ricochet off.